### PR TITLE
Allow specifying Git URL and/or Git branch

### PR DESCRIPTION
**Pull request by stefan-k**
See #4 .

### CODE REVIEW

File: manifests/init.pp
Comment (wiene) [inline manifests/init.pp:24]: What about calling the new variables `cobald_url` and `tardis_url` which default to `undef`? They could be used to specify either the git repository URL or the base URL of Python Package Index. Their meanings depend on the `cobald_repo_type` and the `tardis_repo_type` switches. If they are left to `undef`, the code is fetched from `git+https://github.com/MatterMiners/cobald` and `git+https://github.com/MatterMiners/tardis` in the `git` case and `https://pypi.org/simple` in the `pypi` case.

@stefan-k, @olifre: What do you think?

File: manifests/init.pp
Comment (wiene) [inline manifests/init.pp:21]:
```suggestion
  Enum['pypi', 'git']        $cobald_repo_type           = 'pypi',                     # cobald repository type (PyPI or git)
  Enum['pypi', 'git']        $tardis_repo_type           = 'pypi',                     # tardis repository type (PyPI or git)
```
I think the word "provider" is misleading here. I associate the entity from which one fetches the code with "provider". It also has a special meaning in the Puppet context.

File: manifests/init.pp
Comment (wiene) [inline manifests/init.pp:19]: Setting `cobald_version` or `tardis_version` to `master` or any other branch name will not work, if an installation from PyPi is requested. These cases need to be caught.

Another question is what happens if one leaves `cobald_version` or `tardis_version` to `undef` and chooses to install from git. Probably the best option would be to use the `master` branch. @stefan-k, @olifre: What is your opinion on this?

The comment needs to be updated to reflect the new situation.

**Review by wiene — CHANGES_REQUESTED** [review]
I like the additional flexibility introduced by your proposed change. More detailed thoughts are added in my comments.

Could you please also add a "(closes #4)" to your commit message when updating your PR such that the corresponding issue will be automatically closed after merging. Thanks!

File: manifests/init.pp
Comment (olifre) [reply manifests/init.pp:24]: I think that's a good idea, since also PyPI may use a different URL. Of course, that means that the code installing from PyPI needs to be extended to use that URL, too (and default to `https://pypi.org/simple`). It seems that works via the `index` class parameter of the `puppet-python` classes.

File: manifests/init.pp
Comment (olifre) [reply manifests/init.pp:19]: > Another question is what happens if one leaves `cobald_version` or `tardis_version` to `undef` and chooses to install from git.

Since the current implementation does not specify `Optional`, it can never be set to `undef`. On the other hand, I agree that changing that might be a good idea: `master` or the latest version from `PyPI` seem like sensible defaults.

File: manifests/init.pp
Comment (stefan-k) [reply manifests/init.pp]: Thanks for your suggestions. I named the new variables `cobald_repo_url` and `tardis_repo_url` because `cobald_url` and `tardis_url` are used in `install.pp` and cannot be overwritten (weird puppet...). I chose renaming the exposed parameters instead of the internal variables because I think this is more consistent with `cobald_repo_type` and `tardis_repo_type`.

**Comment by stefan-k** [comment]
Thanks a lot for your comments and suggestions! I tried to address them all. I can squash the commits now for the next review or right before merging, whatever you prefer.

File: manifests/install.pp
Comment (wiene) [inline manifests/install.pp:463]: I think this distinction of cases (`if $cobald_repo_type == 'git' ... else ...`) is not necessary or do I miss anything?

File: manifests/install.pp
Comment (wiene) [inline manifests/install.pp:484]: I think here the distinction of cases is not necessary either.

**Review by wiene — COMMENTED** [review]
@stefan-k: Thanks for your changes. I am mostly happy. I only have two minor comments/questions (which address the same issue).

File: manifests/install.pp
Comment (olifre) [reply manifests/install.pp:463]: I also don't see a reason for the distinction of cases. Dropping it will also remove the syntax `-> if` which is not so nice to read. As a general rule, the most readable practice seems to be to store the differences of cases in attributes defined before the resource or in a hash, and then apply that to the ressource (which is actually exactly what you do with `pip_cobald` and `cobald_url` :wink:  ).

**Review by olifre — CHANGES_REQUESTED** [review]
Indeed, I also don't spot anything apart from the points @wiene has already raised. Thanks for the update!

**Comment by stefan-k** [comment]
Thanks again! I hope this is how you had it in mind.

**Comment by olifre** [comment]
> Thanks again! I hope this is how you had it in mind.

Indeed, and I realize only in the updated version that in fact there was a difference between the Git and PyPI based resources :wink:. So for sure the new compressed version is more readable :+1: . 
@wiene Do you spot anything else? It LGTM so if you don't spot anything, it can be squashed and :rocket: .

**Review by olifre — APPROVED** [review]

**Review by wiene — COMMENTED** [review]
LGTM. Could you please squash your commits to a single commit with a concise message and add "(closes #4)". Then this PR is ready to merge from my point of view.

**Review by wiene — APPROVED** [review]
Perfect, thanks!

**Pull request merged**
